docs/source/en/agents.md

```diff
@@ -0,0 +1,570 @@
+- An attribute `output_type` specifying the type of the ouput. 
+- A `__call__` method which contains the inference code. This is the code we've played with above!
+
+Here's what our class looks like now:
+
+```python
+from transformers import Tool
+from huggingface_hub import list_models
+
+class HFModelDownloadsTool(Tool):
+    name = "model_download_counter"
+    description = (
+        "This is a tool that returns the most downloaded model of a given task on the Hugging Face Hub. "
+        "It returns the name of the checkpoint."
+    )
+
+    inputs = {
+        "task": {
+            "type": str,
+            "description": "the task category (such as text-classification, depth-estimation, etc)",
+        }
+    }
+    output_type = str
+
+    def __call__(self, task: str):
+        model = next(iter(list_models(filter=task, sort="downloads", direction=-1)))
+        return model.id
+```
+
+We now have our tool handy. Save it in a file and import it from your main script. Let's name this file
+`model_downloads.py`, so the resulting import code looks like this:
+
+```python
+from model_downloads import HFModelDownloadsTool
+
+tool = HFModelDownloadsTool()
+```
+
+In order to let others benefit from it and for simpler initialization, we recommend pushing it to the Hub under your 
+namespace. To do so, just call `push_to_hub` on the `tool` variable (be sure to have created the space beforehand, and to be authenticated with a token that has read access!)
+
+```python
+tool.push_to_hub("{your_username}/hf-model-downloads")
+```
+
+You now have your code on the Hub! Let's take a look at the final step, which is to have the agent use it.
+
+### Initializing an agent that uses the tool
+
+We now have our tool that lives on the Hub which can be instantiated as such (change the user name for your tool):
+
+```python
+from transformers import load_tool
+
+model_download_tool = load_tool("m-ric/hf-model-downloads")
+```
+
+In order to use it in the agent, simply pass it to the agent initialization method.
+
+```python
+from transformers import CodeAgent
+
+agent = CodeAgent(llm_engine, tools=[model_download_tool])
+
+agent.run(
+    "Can you give me the name of the model that has the most downloads in the 'text-to-video' task on the Hugging Face Hub?"
+)
+```
+You get the following:
+```text
+==Executing the code below:==
+task = 'text-to-video'
+model_name = model_download_counter(task=task)
+print(f"The most downloaded model in the '{task}' task is '{model_name}'.")
+```
+
+And the output:
+`"The most downloaded model in the 'text-to-video' task is 'ByteDance/AnimateDiff-Lightning'."`
+
+
+
+<Tip>
+
+Depending on the LLM, some are quite brittle and require very exact prompts in order to work well. Having a well-defined
+name and description of the tool is paramount to having it be leveraged by the agent.
+
+</Tip>
+
+### Adding a new tool to an existing agent's toolbox
```

Comment:
Combine this section and the next one into a `### Manage agent toolbox` section. I'll finish reviewing the rest of the doc tomorrow :)